misk/src/main/kotlin/misk/web/actions/ReadinessCheckService.kt

```diff
@@ -0,0 +1,83 @@
+package misk.web.actions
+
+import com.google.common.annotations.VisibleForTesting
+import com.google.common.util.concurrent.AbstractIdleService
+import com.google.common.util.concurrent.ServiceManager
+import misk.healthchecks.HealthCheck
+import misk.healthchecks.HealthStatus
+import misk.tasks.RepeatedTaskQueue
+import misk.tasks.Result
+import misk.tasks.Status
+import misk.web.ReadinessRefreshQueue
+import misk.web.WebConfig
+import wisp.logging.getLogger
+import java.time.Clock
+import java.time.Duration
+import java.time.Instant
+import javax.inject.Inject
+import javax.inject.Provider
+import javax.inject.Singleton
+
+@Singleton
+class ReadinessCheckService @Inject constructor(
+  private val config: WebConfig,
+  private val clock: Clock,
+  private val serviceManagerProvider: Provider<ServiceManager>,
+  @JvmSuppressWildcards private val healthChecks: List<HealthCheck>,
+  @ReadinessRefreshQueue private val taskQueue: RepeatedTaskQueue,
+): AbstractIdleService() {
+  var status: CachedStatus? = null
+    private set
+
+  init {
+    require(config.readiness_refresh_interval_ms < config.readiness_max_age_ms || config.health_port <= 0) {
+      "readiness_refresh_interval_ms must be less than readiness_max_age_ms"
+    }
+  }
+
+  override fun startUp() {
+    logger.info { "starting readiness service" }
+
+    val delay = Duration.ofMillis(config.readiness_refresh_interval_ms.toLong())
+    taskQueue.schedule(delay) {
+      refreshStatuses()
+      Result(Status.OK, delay)
+    }
+  }
+
+  override fun shutDown() {
+    logger.info { "stopping readiness service" }
+
+    status = null
+  }
+
+  @VisibleForTesting
+  internal fun refreshStatuses() {
+    val servicesNotRunning = serviceManagerProvider.get().servicesByState().values().asList()
+      .filterNot { it.isRunning }
+
+    for (service in servicesNotRunning) {
+      logger.info("Service not running: $service")
+    }
+
+    // Don't do healthchecks if services haven't all started. The app isn't in a good state yet,
+    // and a health check could end up triggering random errors that we don't want to flood the
+    // logs with.
+    if (!servicesNotRunning.isEmpty()) return
+
+    val statuses = healthChecks.map { it.status() }
```

Comment:
this will be doing all the individual health checks serially here. that's probably _mostly_ fine, but it is going to mean slower refresh than in parallel.

in the limit, this can mean one slow check can delay the updating of health status for everyone else, but we are bounded by the max age (since we'll go unhealthy regardless if it's slower than that to refresh). 

Updating to parallel execution here would help though still requires each check round to wait for all checks to complete before scheduling another one.

Service Container attempts to address this by having a thread per check to drive its refresh which works to a point (it does have a static limit on that thread pool size which if exceeded then puts you at some risk of queue delay though in practice I don't think anyone has more than 50 health checks)

Reply:
This check has always been serial, and I think thats intentional. When this check was handled by the Action directly it was part of a very limited executor (max 8 threads). Im hesitant to increase the thread consumption of this by a significant or variable amount.

This is happening in the background.

Reply:
FWIW: every healthChecks are running on separate threads, and the [status](https://guava.dev/releases/19.0/api/docs/com/google/common/util/concurrent/Service.html#state()) should return immediately ... which makes me wonder what's the value of this cache 😅 ?

Reply:
oh. I got confused between the `HealthCheck` status and the [Service](https://guava.dev/releases/19.0/api/docs/com/google/common/util/concurrent/Service.html#state()) status

Reply:
I worry this loop can take longer than `readiness_max_age_ms` and take the application down. We should add some tracing to get a sense how much each `HealthCheck` is taking, and do something if we're taking longer than `readiness_max_age_ms`

Reply:
We discussed this in slack. A spot check of all health checks in datadog shows only one instance of > 10s response in the last 2 days, and 2 instances of > 5 seconds. Bumping this to 10 seconds as that case should be an error.